File: 프로그래머스/lv2/12981. 영어 끝말잇기/영어 끝말잇기.py
from collections import defaultdict
def solution(n, words):
    
    answer = []
    word_dict = defaultdict(int)
    word_dict[words[0]] = 1
    i = 1
    
    while i<len(words):
        if words[i] not in word_dict.keys() and words[i-1][-1] == words[i][0]:        
            word_dict[words[i]] = 1
        else:
            answer.append((i%n)+1)
            answer.append((i//n)+1)
            break
            
        i+=1
        
    if len(answer) == 0:
        answer.append(0)
        answer.append(0)
    return answer

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;

class Solution {
    public int[] solution(int n, String[] words) {
        int[] answer = new int[]{0,0};

        HashMap<String,Boolean> hashMap = new HashMap<>();
        
        for(String word : words){
            if(!hashMap.containsKey(word)){
                hashMap.put(word,true);
            }
        }
        for(int i = 0; i < words.length; i++){
            String word = words[i];
            if(hashMap.get(word)){
                hashMap.put(word,false);
            }else{
                answer[0] = i%n + 1;
                answer[1] = i/n + 1;
                return answer;
            }
        }
        
        for(int i = 1;i<words.length;i++){
            String beforeWord = words[i-1];
            String nextWord = words[i];
            
            if(beforeWord.charAt(beforeWord.length()-1) != nextWord.charAt(0)){
                System.out.println(beforeWord);
                System.out.println(nextWord);
                answer[0] = i%n + 1;
                answer[1] = i/n + 1;

                return answer;
            }
        }
        
        
        

        return answer;
    }
}